File: backend/models/grading.py
from datetime import datetime, timezone
from typing import Literal

from pydantic import Field

from .base import JsonModel


GRADING_VERSION = "1.0"


class GradingEntry(JsonModel):
    name: str       # "smog" | "flesch_kincaid" | "dale_chall" | "pemat" | "sam" | "cdc_cci" | "combined"
    target: Literal["before", "after"]
    grade: float    # 0-100 normalized score
    description: str | None = None
    grade_breakdown: dict | None = None
    reasoning: str | None = None


class Grading(JsonModel):
    entries: list[GradingEntry] = Field(default_factory=list)
    enabled: bool = True
    graded_at: str | None = None


_METHOD_REASONING = {
    "smog":           "SMOG (McLaughlin 1969) — counts polysyllabic words; designed for health materials",
    "flesch_kincaid": "Flesch-Kincaid Reading Ease + Grade Level (1975) — sentence length × syllable load",
    "dale_chall":     "Dale-Chall (1948/1995) — difficult words outside the 3,000 familiar-word list",
    "pemat":          "PEMAT (AHRQ 2013) — automated approximation of items 3,8,14,21-22 (understandability) and 27-33 (actionability)",
    "sam":            "SAM (Doak et al. 1996) — automated approximation of content, literacy demand, and layout/typography domains",
    "cdc_cci":        "CDC Clear Communication Index — automated approximation of main message, behavioral recommendations, numbers, and call-to-action items",
}


def build_grading(
    before_score: dict | None, before_text: str | None,
    after_score: dict | None, after_text: str | None,
) -> Grading:
    from utils.scoring_methods import compute_method_scores

    entries = []
    for target, score, text in (
        ("before", before_score, before_text),
        ("after", after_score, after_text),
    ):
        if score is None or text is None:
            continue
        methods = compute_method_scores(text, score["dimensions"])
        for method_name in ("smog", "flesch_kincaid", "dale_chall", "pemat", "sam", "cdc_cci"):
            m = methods[method_name]
            breakdown = {k: v for k, v in m.items() if k != "score"}
            entries.append(GradingEntry(
                name=method_name,
                target=target,
                grade=m["score"],
                grade_breakdown=breakdown,
                reasoning=_METHOD_REASONING[method_name],
            ))
        entries.append(GradingEntry(
            name="combined",
            target=target,
            grade=score["composite"],
            grade_breakdown={
                "grade_estimate": score["grade_estimate"],
                "label":          score["label"],
                "word_count":     score["word_count"],
                "dimensions":     score["dimensions"],
            },
            reasoning=None,
        ))
    return Grading(
        entries=entries,
        enabled=True,
        graded_at=datetime.now(timezone.utc).isoformat(),
    )

File: frontend/src/utils/grading.ts
import type { Grading, GradingEntry } from '../types/envelope';
import type { PatientScore } from '../types/carePlan';

export function patientScoreFromGrading(grading: Grading, target: 'before' | 'after'): PatientScore | null {
  const combined = grading.entries.find(e => e.name === 'combined' && e.target === target);
  if (!combined?.grade_breakdown) return null;
  const bd = combined.grade_breakdown as Record<string, unknown>;
  const dims = bd.dimensions;
  if (!dims) return null;
  return {
    composite:      combined.grade,
    grade_estimate: (bd.grade_estimate as number) ?? 0,
    label:          (bd.label as string) ?? '',
    word_count:     (bd.word_count as number) ?? 0,
    dimensions: Object.fromEntries(
      Object.entries(dims as Record<string, Record<string, unknown>>).map(([k, v]) => [
        k,
        { score: v.score, raw: v.raw, label: v.label, unit: v.unit },
      ])
    ) as PatientScore['dimensions'],
  };
}

export function methodEntriesFromGrading(grading: Grading, target: 'before' | 'after'): GradingEntry[] {
  return grading.entries.filter(e => e.target === target && e.name !== 'combined');
}

export interface MethodGroup {
  name: string;
  label: string;
  before: GradingEntry | undefined;
  after: GradingEntry | undefined;
}

const METHOD_ORDER = ['combined', 'smog', 'flesch_kincaid', 'dale_chall', 'pemat', 'sam', 'cdc_cci'];

const METHOD_LABELS: Record<string, string> = {
  combined:       'Combined Score',
  smog:           'SMOG',
  flesch_kincaid: 'Flesch-Kincaid',
  dale_chall:     'Dale-Chall',
  pemat:          'PEMAT',
  sam:            'SAM',
  cdc_cci:        'CDC Clear Comm.',
};

export function groupGradingEntries(entries: GradingEntry[]): MethodGroup[] {
  const map = new Map<string, MethodGroup>();
  for (const entry of entries) {
    if (!map.has(entry.name)) {
      map.set(entry.name, {
        name: entry.name,
        label: METHOD_LABELS[entry.name] ?? entry.name,
        before: undefined,
        after: undefined,
      });
    }
    const g = map.get(entry.name)!;
    if (entry.target === 'before') g.before = entry;
    else g.after = entry;
  }
  const known = METHOD_ORDER.filter(n => map.has(n)).map(n => map.get(n)!);
  const unknown = [...map.values()].filter(g => !METHOD_ORDER.includes(g.name));
  return [...known, ...unknown];
}

export function combinedScoreLabel(groups: MethodGroup[]): string {
  const combined = groups.find(g => g.name === 'combined');
  if (!combined) return 'Score';
  const before = combined.before ? Math.round(combined.before.grade) : null;
  const after  = combined.after  ? Math.round(combined.after.grade)  : null;
  if (before !== null && after !== null) return `Score (${before} → ${after})`;
  if (after  !== null) return `Score (${after})`;
  if (before !== null) return `Score (${before})`;
  return 'Score';
}
